keel-web/src/main/kotlin/com/netflix/spinnaker/keel/services/ExportService.kt

```diff
@@ -0,0 +1,324 @@
+    }
+  }
+
+  /**
+   * Attempts to extract [SubmittedEnvironment]s from the deploy stages within the pipeline.
+   */
+  private fun deploymentsToEnvironments(
+    pipeline: Pipeline,
+    artifactProviderStage: Stage,
+    artifact: DebianArtifact
+  ): List<SubmittedEnvironment> {
+    val deploys = pipeline.findDownstreamDeploys(artifactProviderStage)
+    val environments = deploys
+      .mapNotNull { deploy ->
+        if (deploy == null || deploy.clusters.isEmpty()) {
+          log.warn("Downstream deploy stage not found after stage ${artifactProviderStage.name} for pipeline ${pipeline.name}")
+          return@mapNotNull null
+        }
+
+        if (deploy.clusters.size > 1) {
+          log.warn("Deploy stage ${deploy.name} for pipeline ${pipeline.name} has more than 1 cluster. Exporting only the first.")
+        }
+
+        // TODO: handle load balancers
+        val cluster = deploy.clusters.first()
+        log.debug("Attempting to build environment for cluster ${cluster.moniker}")
+
+        val provider = cluster.cloudProvider
+        val kind = PROVIDERS_TO_CLUSTER_KINDS[provider] ?: error("Unsupported cluster cloud provider '$provider'")
+        val handler = handlers.supporting(kind)
+        val spec = handler.export(deploy, artifact)
+
+        if (spec == null) {
+          log.warn("No resource exported for kind $kind")
+          return@mapNotNull null
+        }
+
+        val constraints = if (pipeline.hasManualJudgment(deploy)) {
+          log.debug("Adding manual judgment constraint for environment ${cluster.name} based on manual judgment stage in pipeline ${pipeline.name}")
+          setOf(ManualJudgementConstraint())
+        } else {
+          emptySet<Constraint>()
+        }
+
+        val environment = SubmittedEnvironment(
+          name = if (deploys.size > 1) "${cluster.name}-${cluster.region}" else cluster.name,
+          resources = setOf(
+            SubmittedResource(
+              kind = kind,
+              metadata = emptyMap(), // TODO
+              spec = spec
+            )
+          ),
+          constraints = constraints
+        )
+
+        return@mapNotNull environment
+      }
+    return environments
+  }
+
+  /**
+   * Attempts to extract [Constraint]s from the pipeline triggers.
+   */
+  private fun triggersToConstraints(
+    application: String,
+    pipeline: Pipeline,
+    environment: SubmittedEnvironment,
+    pipelinesToEnvironments: Map<Pipeline, List<SubmittedEnvironment>>
+  ): Set<Constraint> {
+    val constraints = mutableSetOf<Constraint>()
+
+    val trigger = pipeline.triggers.also {
+      if (it.size > 1) {
+        log.warn("Pipeline has more than 1 trigger. Using only the first.")
+      }
+    }.firstOrNull()
+
+    if (trigger == null || !trigger.enabled) {
+      // if there's no trigger, the pipeline is triggered manually, i.e. the equivalent of a manual judgment
```

Comment:
In most simple cases I've seen, production environments should have two constraints: depends-on (testing), and manual-judgement.  It looks like from this if/else block, it'd only be configured to have the one constraint.

Reply:
Good point. I'm already having to find the upstream environment below for pipeline triggers, but could do something similar for pipelines with no triggers, starting from their first deploy stage.